CenterForOpenScience/osf.io
# [Feature] Index Files

## Purpose
Improve search results using the information in files by indexing the files stored in a public project's osf storage for use when querying.

## Dependencies
* Uses [elasticsearch-mapper-attachments](https://github.com/elastic/elasticsearch-mapper-attachments) plugin to extract.

## Functionality
* When searching, projects with files in osfstorage that match a given query are returned in results.
* Edits to files are reflected in search results.
* Moving or coping a file into a different project causes that project to appear in the same search results.
* Only indexes documents of type rtf, pdf, docx, md, and txt.
* Only indexes documents in osf storage.
* Files are added to elasticsearch when a project is made public.
* Files are removed from elasticsearch when a project is made private.
* Improve search result ordering
  * Reduce weight of a file's content for relevance calculation.
  * Increase weight given to contributors and tags to prevent overshadowing by files.

#### Text Indexed
* Project
  * Title (4)
  * Contributors (3)
  * Tags (2)
  * Description (1.2)
 * Wiki pages
* File
  * Name (0.5)
  * Content (0.01)
* User
  * Fullname
  * Names
  * all jobs (0.125)
  * job title
  * job
  * school
  * all schools (0.125)
  * degree

## Testing Considerations
* Ensure multiple moves, deletes, and copies handled correctly.
* Ensure moves between projects and components with different privacy settings are handled correctly.
* Ensure search results have a reasonable ordering.
